cmd/backfill-redis/main.go

```diff
@@ -0,0 +1,144 @@
+	redisPort    = flag.String("port", "", "Port to Redis application")
+	startIndex   = flag.Int("start", -1, "First index to backfill")
+	endIndex     = flag.Int("end", -1, "Last index to backfill")
+	rekorAddress = flag.String("rekor-address", "", "Address for Rekor, e.g. https://rekor.sigstore.dev")
+)
+
+func main() {
+	flag.Parse()
+
+	if *redisAddress == "" {
+		log.Fatal("address must be set")
+	}
+	if *redisPort == "" {
+		log.Fatal("port must be set")
+	}
+	if *startIndex == -1 {
+		log.Fatal("start must be set to >=0")
+	}
+	if *endIndex == -1 {
+		log.Fatal("end must be set to >=0")
+	}
+	if *rekorAddress == "" {
+		log.Fatal("rekor-address must be set")
+	}
+
+	cfg := radix.PoolConfig{}
+	redisClient, err := cfg.New(context.Background(), "tcp", fmt.Sprintf("%s:%s", *redisAddress, *redisPort))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	rekorClient, err := client.GetRekorClient(*rekorAddress)
+	if err != nil {
+		log.Fatalf("creating rekor client: %v", err)
+	}
+
+	for i := *startIndex; i <= *endIndex; i++ {
+		params := entries.NewGetLogEntryByIndexParamsWithContext(context.Background())
+		params.SetLogIndex(int64(i))
+		resp, err := rekorClient.Entries.GetLogEntryByIndex(params)
+		if err != nil {
+			log.Fatalf("retrieving log uuid by index: %v", err)
+		}
+		success := true
+		for uuid, entry := range resp.Payload {
+			// uuid is the global UUID - tree ID and entry UUID
+			e, _, _, err := unmarshalEntryImpl(entry.Body.(string))
+			if err != nil {
+				fmt.Printf("error unmarshalling entry for %s: %v\n", uuid, err)
+				success = false
+				continue
+			}
+			keys, err := e.IndexKeys()
+			if err != nil {
+				fmt.Printf("error building index keys for %s: %v\n", uuid, err)
+				success = false
+				continue
+			}
+			for _, key := range keys {
+				if err := addToIndex(context.Background(), redisClient, key, uuid); err != nil {
+					success = false
+					fmt.Printf("error inserting UUID %s with key %s: %v\n", uuid, key, err)
+				}
+				fmt.Printf("Uploaded Redis entry %s, index %d, key %s\n", uuid, i, key)
+			}
+		}
+		if success {
+			fmt.Printf("Completed log index %d\n", i)
+		} else {
+			fmt.Printf("Errors with log index %d\n", i)
+		}
+	}
+}
+
+// unmarshalEntryImpl decodes the base64-encoded entry to a specific entry type (types.EntryImpl).
+// from cosign
+func unmarshalEntryImpl(e string) (types.EntryImpl, string, string, error) {
+	b, err := base64.StdEncoding.DecodeString(e)
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	pe, err := models.UnmarshalProposedEntry(bytes.NewReader(b), runtime.JSONConsumer())
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	entry, err := types.UnmarshalEntry(pe)
+	if err != nil {
+		return nil, "", "", err
+	}
+	return entry, pe.Kind(), entry.APIVersion(), nil
+}
+
+func addToIndex(ctx context.Context, redisClient radix.Client, key, value string) error {
+	return redisClient.Do(ctx, radix.Cmd(nil, "LPUSH", key, value))
```

Comment:
a potential optimization:  `LPUSH key a b c` is equivalent to `LPUSH key c`, `LPUSH key b`, and `LPUSH key a` per https://redis.io/commands/lpush/

Reply:
I think we'd need the inverse, because for one entry, we have many keys mapping to one value, rather than one key mapping to many values.

Reply:
Actually, just realized, if we have the wrong range and we accidentally backfill an existing index, then it'll push a second duplicate UUID into the list. Not sure if this causes any issues.

To be safe, I added a remove before the push.

Reply:
Do you know why we are using lists and not sets?

Reply:
I was wondering about batch mode as well, but wasn't sure how the errors are reported / handled for if things fail.

Reply:
I couldn't find any ways to batch this using the available commands. Performance shouldn't be an issue either for a few thousands entries.

Reply:
So for our test cases in staging then, I reckon we'll do something like this?

Say we have existing indices, [0,12] (these are obvs not right, but just for illustrative purposes)

Delete 8-10

Run the script from 5-12

And make sure things work still?

Reply:
I'd narrow down the window of extra indices. Let's say we delete 8-10, and run it on 7-11. Before, we should check all key-value pairs for 7 and 11, and make sure that they remain the same after running